crux_macros/src/effect.rs

```diff
@@ -0,0 +1,235 @@
+use convert_case::{Case, Casing};
+use darling::{ast, util, FromDeriveInput, FromField, FromMeta, ToTokens};
+use proc_macro2::TokenStream;
+use quote::quote;
+use syn::{DeriveInput, GenericArgument, Ident, PathArguments, Type};
+
+#[derive(FromDeriveInput, Debug)]
+#[darling(attributes(effect), supports(struct_named))]
+struct EffectStructReceiver {
+    ident: Ident,
+    name: Option<Type>,
+    app: Option<Type>,
+    data: ast::Data<util::Ignored, EffectFieldReceiver>,
+}
+
+#[derive(FromField, Debug)]
+#[darling(attributes(effect))]
+pub struct EffectFieldReceiver {
+    ident: Option<Ident>,
+    ty: Type,
+}
+
+impl ToTokens for EffectStructReceiver {
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        let ident = &self.ident;
+
+        let name = match self.name {
+            Some(ref name) => quote!(#name),
+            None => {
+                let x = Type::from_string("Effect").unwrap();
+                quote!(#x)
+            }
+        };
+
+        let app = match self.app {
+            Some(ref app) => quote!(#app),
+            None => {
+                let x = Type::from_string("App").unwrap();
+                quote!(#x)
+            }
+        };
+
+        let fields = self
+            .data
+            .as_ref()
+            .take_struct()
+            .expect("Should never be enum")
+            .fields;
+
+        let event = extract_event_type(&fields.first().as_ref().unwrap().ty);
+        let event = fields
+            .iter()
+            .map(|f| extract_event_type(&f.ty))
+            .all(|t| quote!(#t).to_string() == quote!(#event).to_string())
+            .then_some(event)
+            .expect("all fields should be generic over the same event type");
+
+        let (variants, fields): (Vec<_>, Vec<_>) = fields
+            .into_iter()
+            .map(|field| {
+                let (field_name, variant) = field
+                    .ident
+                    .as_ref()
+                    .map(|snake| {
+                        let pascal =
+                            Ident::new(&snake.to_string().to_case(Case::Pascal), snake.span());
+                        (quote!(#snake), quote!(#pascal))
+                    })
+                    .expect("We already told darling we're on a struct with named fields");
+
+                let ty = &field.ty;
+                (
+                    quote! { #variant(<#ty as ::crux_core::capability::Capability<#event>>::Operation) },
+                    // TODO: Make this use the actual type, not #variant
```

Comment:
Spotted this - we probably want to take the name out of ty rather than using `#variant::new` here - have found in the past that you don't always get the correct output when you roundtrip through snakecase->pascalcase

Reply:
I updated this to split the type, instead of converting from snake to pascal.